docs/testing/e2e-framework-design-dimensions.md

```diff
@@ -0,0 +1,396 @@
+| 5         | Kubernetes versions          |
+| 10        | Node OSes                    |
+| 5         | Node OS versions             |
+| 5         | Packages controller versions |
+| 5         | Package versions             |
+| 2         | Architectures                |
+| 2         | Self vs Remotely managed     |
+| 2         | Bundle version               |
+| 6,250,000 | Total permutations           |
+
+With 6,250,000 permutations, running on 32 CPUs at a rate of one test per CPU per minute, a complete test run would take ≈ 4½ months.
+
+A more middle-of-the-road estimate might be:
+
+|       |                              |
+|-------|------------------------------|
+| 5     | Providers                    |
+| 1.6   | Provider versions            |
+| 2     | CLI versions                 |
+| 4     | Kubernetes versions          |
+| 3     | Node OSes                    |
+| 2     | Node OS versions             |
+| 2     | Packages controller versions |
+| 2     | Package versions             |
+| 1     | Architectures                |
+| 1     | Self vs Remotely managed     |
+| 1     | Bundle version               |
+| 1,536 | Total permutations           |
+
+With this far more tractable scenario, a complete run would take forty-eight minutes. Of course this still assumes 32 CPUs available and able to run any test (none of which take more than a minute from beginning to end).
+
+<style>
+.markdown-body #worst-case-cardinality ~ table tbody tr,
+.markdown-body #worst-case-cardinality ~ table tbody tr.even td,
+.markdown-body #worst-case-cardinality ~ table tbody tr.even th,
+.markdown-body #worst-case-cardinality ~ table tbody tr.odd td,
+.markdown-body #worst-case-cardinality ~ table tbody tr.odd th {
+    background-color: inherit;
+    border: none;
+}
+
+.markdown-body #worst-case-cardinality ~ table tbody tr td {
+    padding: 0 0.5em;
+}
+
+.markdown-body #worst-case-cardinality ~ table td:first-child {
+    text-align: right;
+}
+
+.markdown-body #worst-case-cardinality ~ table tbody tr:last-child {
+    border-top: 1px solid #c9d1d9;
+
+}
+
+.markdown-body #worst-case-cardinality ~ table tbody tr:last-child td {
+    padding-top: 5px;
+}
+</style>
+
+
+<a id="orgc40bdd7"></a>
+
+## Implementation Design Ideas
+
+
+<a id="orgf548cac"></a>
+
+### Go tags via comments
+
+Tests could be marked with go tags, and included or excluded based on their matching a run specification:
+
+```go
+// e2e:providers=["vSphere", "Docker"] 
+// e2e:nodeOS!=["Red Hat"] 
+func TestSomePackage(t *testing.T) { 
+	assert.True(t, true) 
+} 
+```
+
+This approach makes tests easy to identify, and is simple for test writers to implement. However there's added complexity in writing a tool to read, parse, and react to the go tags. The tags themselves are a bit magical, and will depend on documentation for discovery.
+
+
+<a id="orgdbecf99"></a>
+
+### Test functions to build state
+
+```go 
+func TestSomePackage(t \*testing.T) { 
```

Comment:
This could be a bit confusing if there are test scenarios that change between providers. Example Harbor would test differently in a dynamic provisioning environment vs not.

Reply:
Presumably, the test writer would have access to some structure in which they could look up the current test dimensions under which the test was executing, and take the appropriate actions based on that.